File: scripts/day7-part2-optimized.js
const fs = require("fs");

function solvePart2(grid) {
  const rows = grid.length;
  const cols = grid[0].length;

  // Find the starting position 'S'
  let startCol = grid[0].indexOf("S");

  // Track timelines per state
  let currentStates = new Map([[`0,${startCol}`, 1]]);

  for (let row = 0; row < rows; row++) {
    const nextStates = new Map();

    for (const [stateKey, count] of currentStates) {
      const [r, c] = stateKey.split(",").map(Number);
      if (r !== row) continue;

      const cell = grid[r][c];

      if (cell === "." || cell === "S") {
        const nextKey = `${r + 1},${c}`;
        nextStates.set(nextKey, (nextStates.get(nextKey) || 0) + count);
      } else if (cell === "^") {
        if (c > 0) {
          const leftKey = `${r + 1},${c - 1}`;
          nextStates.set(leftKey, (nextStates.get(leftKey) || 0) + count);
        }
        if (c < cols - 1) {
          const rightKey = `${r + 1},${c + 1}`;
          nextStates.set(rightKey, (nextStates.get(rightKey) || 0) + count);
        }
      }
    }

    currentStates = nextStates;
  }

  let totalTimelines = 0;
  for (const count of currentStates.values()) {
    totalTimelines += count;
  }

  return totalTimelines;
}

const input = fs
  .readFileSync("../inputs/day-7-input.txt", "utf-8")
  .trim()
  .split(/\r?\n/);

const iterations = 1;
const startTime = process.hrtime.bigint();

let result;
for (let i = 0; i < iterations; i++) {
  result = solvePart2(input);
}

const endTime = process.hrtime.bigint();
const totalNanos = Number(endTime - startTime);
const avgMicros = (totalNanos / iterations / 1000).toFixed(3);

console.log(`JavaScript (Day 7 Part 2 - Optimized)`);
console.log(`Answer: ${result}`);
console.log(`Average time: ${avgMicros} microseconds (${iterations} iterations)`);
console.log(`Total time: ${(totalNanos / 1000000).toFixed(2)} ms`);

File: scripts/day7-part2-optimized.py
import time

def solve_part2(grid):
    rows = len(grid)
    cols = len(grid[0])
    
    # Find the starting position 'S'
    start_col = grid[0].index('S')
    
    # Track timelines per state
    current_states = {(0, start_col): 1}
    
    for row in range(rows):
        next_states = {}
        
        for (r, c), count in current_states.items():
            if r != row:
                continue
            
            cell = grid[r][c]
            
            if cell == '.' or cell == 'S':
                next_key = (r + 1, c)
                next_states[next_key] = next_states.get(next_key, 0) + count
            elif cell == '^':
                if c > 0:
                    left_key = (r + 1, c - 1)
                    next_states[left_key] = next_states.get(left_key, 0) + count
                if c < cols - 1:
                    right_key = (r + 1, c + 1)
                    next_states[right_key] = next_states.get(right_key, 0) + count
        
        current_states = next_states
    
    total_timelines = sum(current_states.values())
    
    return total_timelines

# Read input
with open("../inputs/day-7-input.txt", "r") as f:
    input_data = f.read().strip().split('\n')

# Benchmark
iterations = 1
start_time = time.perf_counter()

for _ in range(iterations):
    result = solve_part2(input_data)

end_time = time.perf_counter()
total_time = end_time - start_time
avg_time_us = (total_time / iterations) * 1_000_000

print(f"Python (Day 7 Part 2 - Optimized)")
print(f"Answer: {result}")
print(f"Average time: {avg_time_us:.3f} microseconds ({iterations} iterations)")
print(f"Total time: {total_time * 1000:.2f} ms")
